keps/sig-storage/1472-storage-capacity-tracking/README.md

```diff
@@ -767,24 +830,35 @@ enhancement:
       - kube-scheduler
 
 * **Does enabling the feature change any default behavior?**
-  No changes for existing storage driver deployments.
+
+  Enabling it only in kube-scheduler and api-server and not any of the
+  running CSI drivers causes no changes. Everything continues as
+  before because no `CSIStorageCapacity` objects are created and
+  kube-scheduler does not wait for any.
+
+  That changes once the feature is enabled in a CSI driver. Then pod
+  scheduling becomes more likely to pick suitable nodes. This happens
+  automatically, without having to change application deployments.
 
 * **Can the feature be disabled once it has been enabled (i.e. can we rollback
   the enablement)?**
   Yes.
 
-  When disabling support in the API server, the new object type and
-  thus all objects will disappear together with the new flag in
-  `CSIDriver`, which will then cause kube-scheduler to revert to
-  provisioning without capacity information. However, the new objects
-  will still remain in the etcd database.
+  In Kubernetes 1.19 and 1.20, registration of the
+  `CSIStorageCapacity` type was controlled by the feature gate. In
+  1.21, the type will always be enabled in the v1beta1 API
+  group. Depending on the combination of Kubernetes release and
+  feature gate, the type will be disabled. However, any existing
+  objects will still remain in the etcd database, they just won't be
+  visible.
 
-  When disabling it in the scheduler, provisioning directly reverts
-  to the previous behavior.
+  When the type is disabled, external-provisioner will be unable to update
+  objects: this needs to be treated with exponential backoff just like other
+  communication issues with the API server.
 
-  External-provisioner will be unable to update objects: this needs to
-  be treated with exponential backoff just like other communication
-  issues with the API server.
+  The new flag in `CSIDriver` will always be removed when disabling the
+  feature gate or when rolling back to a release older than 1.19. In that
```

Comment:
re feature-gate disablement:
If the field is set and the FG was disabled, this shouldn't automatically delete the field - we should only drop the field if someone is trying to set it, but not if that already exists (unless someone explicitly removes it).

It's basically this pattern of "if FD disabled and field not in use", e.g.:
https://github.com/kubernetes/kubernetes/blob/074a5177209c3605daf11cfe6f477fddce0d55c0/pkg/api/pod/util.go#L450

Reply:
Again, this is not how any of the other fields in storage objects work. That's what was agreed upon with the API reviewers. I am not sure why it is being done, but I am reluctant to change it without a wider discussion.

Can we merge the KEP as-is and I bring it up when doing the API change?

Reply:
One of the reasons for having PRR is also to achieve predictibility and consistency across the system. Especially for Beta and GA features.
If we're doing something differently then anywhere else, I would like to understand the clear reasons for that and have that documented. If you write "I am not sure why it is being done", I would really like to hear API approvers voice now.

@liggitt - can you please chime in?

Reply:
Allowing existing data to be preserved on update is required. This ensures that an n-1 server where the feature gate is not yet enabled by default will not cause data loss in a skewed multi-server cluster during upgrade. Please update that detail here to avoid confusion before merge.

Reply:
> this is not how any of the other fields in storage objects work

Do you have pointers to those? If we are dropping field values on update without regard for existing persisted data, we'll need to open issues to resolve that.

For reference, the ephemeral volume field is following the expected pattern correctly:

https://github.com/kubernetes/kubernetes/blob/074a5177209c3605daf11cfe6f477fddce0d55c0/pkg/api/pod/util.go#L568-L576

Reply:
Let me rephrase. I know why the behavior makes sense to *me* but I don't know why storage APIs seem to handle this differently than the rest of the system or why API reviewers have not pointed this out earlier.

To me it makes sense to hide the field when the feature is turned off in the apiserver because it provides a central kill switch if things go wrong:
- the scheduler reverts to scheduling without storage capacity: this can of course also be achieved by reconfiguring the kube-scheduler, but it's one additional change
- attempts to deploy a CSI driver with the feature enabled in `CSIDriver` will fail, which is a clear indicator for "this feature is not supported in the cluster": this might be also possible by doing the validation differently (not sure)

Reply:
I disagree - scheduler should revert by just switching of feature gate.

And this generated a significant issue that:
- I create an object with a field set
- FG gets disabled (and that silently drops the field)
- FG gets reenabled, and the object I created with a field set has no longer the field set.

I don't know why this wasn't caught in the review, but that sounds like an omission. I don't think that justifies leaving that for longer. We should fix that (and do the same thing we're doing everywhere else).

Reply:
> Do you have pointers to those? If we are dropping field values on update without regard for existing persisted data, we'll need to open issues to resolve that.

This is the commit which added the additional field: https://github.com/kubernetes/kubernetes/commit/1089954fa6cc3e20ce14fad3df010403808c08a0#diff-1918b63c53bdbabb45108a29e45667799fd5674bc5e04b1dd01c24131a79308f

I was doing the same as for `VolumeLifecycleModes`.  If I read the code in `pkg/registry/storage/csidriver/strategy.go` right, updating a `CSIDriver` object while the feature is disabled preserves the field. I think that addresses the concern about data loss.

But should that field be sent to clients when they ask for a v1 CSIDriver object and the apiserver has the feature disabled? My understanding is that @wojtek-t wants it to be sent whereas the current implementation of that and other fields doesn't do it.

Reply:
... or at least I think it doesn't. Let me double-check that.

Reply:
I was wrong, sorry for the confusion. "kubectl get" doesn't show the field after disabling the feature in the apiserver, but it's still there in the actual object. "kubect get -o yaml" shows it.

That means I need to update the KEP to reflect the implementation, but the implementation itself should be okay.

Reply:
I've pushed a commit with the KEP change. Let's keep it as a separate commit to document in eternity how confused I was about this... :sweat_smile: